README.md
Desctable
================

[![Travis-CI Build
Status](https://travis-ci.org/MaximeWack/desctable.svg?branch=master)](https://travis-ci.org/MaximeWack/desctable)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/desctable)](https://cran.r-project.org/package=desctable)
[![CRAN RStudio mirror
downloads](http://cranlogs.r-pkg.org/badges/desctable)](https://www.r-pkg.org:443/pkg/desctable)

# Introduction

Desctable is a comprehensive descriptive and comparative tables
generator for R.

Every person doing data analysis has to create tables for descriptive
summaries of data (a.k.a. Table.1), or comparative tables.

Many packages, such as the aptly named **tableone**, address this issue.
However, they often include hard-coded behaviors, have outputs not
easily manipulable with standard R tools, or their syntax are
out-of-style (e.g. the argument order makes them difficult to use with
the pipe (`%>%`)).

Enter **desctable**, a package built with the following objectives in
mind:

-   generate descriptive and comparative statistics tables with nesting
-   keep the syntax as simple as possible
-   have good reasonable defaults
-   be entirely customizable, using standard R tools and functions
-   produce the simplest (as a data structure) output possible
-   provide helpers for different outputs
-   integrate with “modern” R usage, and the **tidyverse** set of tools
-   apply functional paradigms

# Installation

Install from CRAN with

    install.packages("desctable")

or install the development version from github with

    devtools::install_github("maximewack/desctable")

# Loading

``` r
library(desctable)
```

It is recommended to read this manual through its vignette:

``` r
vignette("desctable")
```

------------------------------------------------------------------------

# Descriptive tables

## Simple usage

**desctable** uses and exports the pipe (`%>%`) operator (from packages
**magrittr** and **dplyr** fame), though it is not mandatory to use it.

The single interface to the package is its eponymous `desctable`
function.

When used on a data.frame, it returns a descriptive table:

``` r
iris %>%
  desctable()
```

    ##                         N        %     Mean        sd  Med IQR
    ## 1        Sepal.Length 150       NA       NA        NA 5.80 1.3
    ## 2         Sepal.Width 150       NA 3.057333 0.4358663 3.00 0.5
    ## 3        Petal.Length 150       NA       NA        NA 4.35 3.5
    ## 4         Petal.Width 150       NA       NA        NA 1.30 1.5
    ## 5             Species 150       NA       NA        NA   NA  NA
    ## 6     Species: setosa  50 33.33333       NA        NA   NA  NA
    ## 7 Species: versicolor  50 33.33333       NA        NA   NA  NA
    ## 8  Species: virginica  50 33.33333       NA        NA   NA  NA

``` r
desctable(mtcars)
```

    ##          N      Mean        sd     Med       IQR
    ## 1   mpg 32 20.090625 6.0269481  19.200   7.37500
    ## 2   cyl 32        NA        NA   6.000   4.00000
    ## 3  disp 32        NA        NA 196.300 205.17500
    ## 4    hp 32        NA        NA 123.000  83.50000
    ## 5  drat 32  3.596563 0.5346787   3.695   0.84000
    ## 6    wt 32        NA        NA   3.325   1.02875
    ## 7  qsec 32 17.848750 1.7869432  17.710   2.00750
    ## 8    vs 32        NA        NA   0.000   1.00000
    ## 9    am 32        NA        NA   0.000   1.00000
    ## 10 gear 32        NA        NA   4.000   1.00000
    ## 11 carb 32        NA        NA   2.000   2.00000

<br>

As you can see with these two examples, `desctable` describes every
variable, with individual levels for factors. It picks statistical
functions depending on the type and distribution of the variables in the
data, and applies those statistical functions only on the relevant
variables.

## Output

The object produced by `desctable` is in fact a list of data.frames,
with a “desctable” class.  
Methods for reduction to a simple dataframe (`as.data.frame`,
automatically used for printing), conversion to markdown (`pander`), and
interactive html output with **DT** (`datatable`) are provided:

``` r
iris %>%
  desctable() %>%
  pander()
```

|                | N   | %   | Mean | sd   | Med | IQR |
|:---------------|:----|:----|:-----|:-----|:----|:----|
| Sepal.Length   | 150 |     |      |      | 5.8 | 1.3 |
| Sepal.Width    | 150 |     | 3.1  | 0.44 | 3   | 0.5 |
| Petal.Length   | 150 |     |      |      | 4.3 | 3.5 |
| Petal.Width    | 150 |     |      |      | 1.3 | 1.5 |
| **Species**    | 150 |     |      |      |     |     |
|     setosa     | 50  | 33  |      |      |     |     |
|     versicolor | 50  | 33  |      |      |     |     |
|     virginica  | 50  | 33  |      |      |     |     |

<br>

To use `pander` you need to load the package yourself.

Calls to `pander` and `datatable` with “regular” dataframes will not be
affected by the defaults used in the package, and you can modify these
defaults for **desctable** objects.

The `datatable` wrapper function for desctable objects comes with some
default options and formatting such as freezing the row names and table
header, export buttons, and rounding of values. Both `pander` and
`datatable` wrapper take a *digits* argument to set the number of
decimals to show. (`pander` uses the *digits*, *justify* and *missing*
arguments of `pandoc.table`, whereas `datatable` calls `prettyNum` with
the `digits` parameter, and removes `NA` values. You can set
`digits = NULL` if you want the full table and format it yourself)

Subsequent outputs in this README will use **pander**.

## Advanced usage

`desctable` automatically chooses statistical functions if none is
provided, using the following algorithm:

-   always show N
-   if there are factors, show %
-   if there are normally distributed variables, show Mean and SD
-   if there are non-normally distributed variables, show Median and IQR

For each variable in the table, compute the relevant statistical
functions in that list (non-applicable functions will safely return
`NA`).

You can specify the statistical functions yourself with the *stats*
argument. This argument can either be:

-   a function for automatic selection of appropriate statistical
    functions, depending on the data
-   a named list of functions/formulas

The functions/formulas leverage the **tidyverse** way of working with
anonymous functions, i.e.:

If a *function*, is is used as is. If a *formula*, e.g. ‘\~ .x + 1’ or
`~ . + 1`, it is converted to a function. There are three ways to refer
to the arguments:

-   For a single argument function, use ‘.’
-   For a two argument function, use ‘.x’ and ‘.y’
-   For more arguments, use ‘..1’, ‘..2’, ‘..3’ etc

This syntax allows you to create very compact anonymous functions, and
is the same as in the `map` family of functions from **purrr**.

**Conditional formulas (`condition ~ if_T | if F`) from previous
versions are no longer supported!**

### Automatic function

The default value for the *stats* argument is `stats_auto`, provided in
the package.

Several other “automatic statistical functions” are defined in this
package: `stats_auto`, `stats_default`, `stats_normal`,
`stats_nonnormal`.

You can also provide your own automatic function, which needs to

-   accept a dataframe as its argument (whether to use this dataframe or
    not in the function is your choice), and
-   return a named list of statistical functions to use, as defined in
    the subsequent paragraphs.

``` r
# Strictly equivalent to iris %>% desctable() %>% pander()
iris %>%
  desctable(stats = stats_auto) %>%
  pander()
```

|                | N   | %   | Mean | sd   | Med | IQR |
|:---------------|:----|:----|:-----|:-----|:----|:----|
| Sepal.Length   | 150 |     |      |      | 5.8 | 1.3 |
| Sepal.Width    | 150 |     | 3.1  | 0.44 | 3   | 0.5 |
| Petal.Length   | 150 |     |      |      | 4.3 | 3.5 |
| Petal.Width    | 150 |     |      |      | 1.3 | 1.5 |
| **Species**    | 150 |     |      |      |     |     |
|     setosa     | 50  | 33  |      |      |     |     |
|     versicolor | 50  | 33  |      |      |     |     |
|     virginica  | 50  | 33  |      |      |     |     |

<br>

For reference, here is the body of the `stats_auto` function in the
package:

    ## function (data) 
    ## {
    ##     shapiro <- data %>% Filter(f = is.numeric) %>% lapply(is.normal) %>% 
    ##         unlist()
    ##     if (length(shapiro) == 0) {
    ##         normal <- F
    ##         nonnormal <- F
    ##     }
    ##     else {
    ##         normal <- any(shapiro)
    ##         nonnormal <- any(!shapiro)
    ##     }
    ##     fact <- data %>% lapply(is.factor) %>% unlist() %>% any()
    ##     if (fact & normal & !nonnormal) 
    ##         stats_normal(data)
    ##     else if (fact & !normal & nonnormal) 
    ##         stats_nonnormal(data)
    ##     else if (fact & !normal & !nonnormal) 
    ##         list(N = length, `%` = percent)
    ##     else if (!fact & normal & nonnormal) 
    ##         list(N = length, Mean = ~if (is.normal(.)) mean(.), sd = ~if (is.normal(.)) sd(.), 
    ##             Med = stats::median, IQR = ~if (!is.factor(.)) IQR(.))
    ##     else if (!fact & normal & !nonnormal) 
    ##         list(N = length, Mean = mean, sd = stats::sd)
    ##     else if (!fact & !normal & nonnormal) 
    ##         list(N = length, Med = stats::median, IQR = IQR)
    ##     else stats_default(data)
    ## }
    ## <bytecode: 0x56144fe581a8>
    ## <environment: namespace:desctable>

<br>

### Statistical functions

Statistical functions can be **any** function defined in R that you want
to use, such as `length` or `mean`.

The only condition is that they return a single numerical value. One
exception is when they return a vector of length `1 + nlevels(x)` when
applied to factors, as is needed for the `percent` function.

As mentioned above, they need to be used inside a **named list**, such
as

``` r
mtcars %>%
  desctable(stats = list("N" = length, "Mean" = mean, "SD" = sd)) %>%
  pander()
```

|      | N   | Mean | SD   |
|:-----|:----|:-----|:-----|
| mpg  | 32  | 20   | 6    |
| cyl  | 32  | 6.2  | 1.8  |
| disp | 32  | 231  | 124  |
| hp   | 32  | 147  | 69   |
| drat | 32  | 3.6  | 0.53 |
| wt   | 32  | 3.2  | 0.98 |
| qsec | 32  | 18   | 1.8  |
| vs   | 32  | 0.44 | 0.5  |
| am   | 32  | 0.41 | 0.5  |
| gear | 32  | 3.7  | 0.74 |
| carb | 32  | 2.8  | 1.6  |

<br>

The names will be used as column headers in the resulting table, and the
functions will be applied safely on the variables (errors return `NA`,
and for factors the function will be used on individual levels).

Several convenience functions are included in this package.

-   `percent`, which prints percentages of levels in a factor
-   `IQR`, which re-implements `stats::IQR` but works better with `NA`
    values
-   `is.normal`, which tests for normality using the following method:
    `length(na.omit(x)) > 30 & shapiro.test(x)$p.value > .1`

Be aware that **all functions will be used on variables stripped of
their `NA` values!** This is necessary for most statistical functions to
be useful, and makes **N** (`length`) show only the number of
observations in the dataset for each variable.

### Labels

It is often the case that variable names are not “pretty” enough to be
used as-is in a table.  
Although you could still edit the variable labels in the table
afterwards using sub-setting or string replacement functions, we provide
a facility for this using the **labels** argument.

The **labels** argument is a named character vector associating variable
names and labels.  
You don’t need to provide labels for all the variables, and extra labels
will be silently discarded. This allows you to define a “global” labels
vector and use it for multiple tables even after variable selections.

``` r
mtlabels <- c(mpg  = "Miles/(US) gallon",
              cyl  = "Number of cylinders",
              disp = "Displacement (cu.in.)",
              hp   = "Gross horsepower",
              drat = "Rear axle ratio",
              wt   = "Weight (1000 lbs)",
              qsec = "¼ mile time",
              vs   = "V/S",
              am   = "Transmission",
              gear = "Number of forward gears",
              carb = "Number of carburetors")

mtcars %>%
  dplyr::mutate(am = factor(am, labels = c("Automatic", "Manual"))) %>%
  desctable(labels = mtlabels) %>%
  pander()
```

|                         | N   | %   | Mean | sd   | Med | IQR  |
|:------------------------|:----|:----|:-----|:-----|:----|:-----|
| Miles/(US) gallon       | 32  |     | 20   | 6    | 19  | 7.4  |
| Number of cylinders     | 32  |     |      |      | 6   | 4    |
| Displacement (cu.in.)   | 32  |     |      |      | 196 | 205  |
| Gross horsepower        | 32  |     |      |      | 123 | 84   |
| Rear axle ratio         | 32  |     | 3.6  | 0.53 | 3.7 | 0.84 |
| Weight (1000 lbs)       | 32  |     |      |      | 3.3 | 1    |
| ¼ mile time             | 32  |     | 18   | 1.8  | 18  | 2    |
| V/S                     | 32  |     |      |      | 0   | 1    |
| **Transmission**        | 32  |     |      |      |     |      |
|     Automatic           | 19  | 59  |      |      |     |      |
|     Manual              | 13  | 41  |      |      |     |      |
| Number of forward gears | 32  |     |      |      | 4   | 1    |
| Number of carburetors   | 32  |     |      |      | 2   | 2    |

<br>

------------------------------------------------------------------------

# Comparative tables

## Simple usage

Creating a comparative table (between groups defined by a factor) using
`desctable` is as easy as creating a descriptive table.

It leverages the `group_by` function from **dplyr**:

``` r
iris %>%
  group_by(Species) %>%
  desctable() -> iris_by_Species

iris_by_Species
```

    ##                Species: setosa (n=50) / N Species: setosa (n=50) / Mean
    ## 1 Sepal.Length                         50                         5.006
    ## 2  Sepal.Width                         50                         3.428
    ## 3 Petal.Length                         50                            NA
    ## 4  Petal.Width                         50                            NA
    ##   Species: setosa (n=50) / sd Species: setosa (n=50) / Med
    ## 1                   0.3524897                          5.0
    ## 2                   0.3790644                          3.4
    ## 3                          NA                          1.5
    ## 4                          NA                          0.2
    ##   Species: setosa (n=50) / IQR Species: versicolor (n=50) / N
    ## 1                        0.400                             50
    ## 2                        0.475                             50
    ## 3                        0.175                             50
    ## 4                        0.100                             50
    ##   Species: versicolor (n=50) / Mean Species: versicolor (n=50) / sd
    ## 1                             5.936                       0.5161711
    ## 2                             2.770                       0.3137983
    ## 3                             4.260                       0.4699110
    ## 4                                NA                              NA
    ##   Species: versicolor (n=50) / Med Species: versicolor (n=50) / IQR
    ## 1                             5.90                            0.700
    ## 2                             2.80                            0.475
    ## 3                             4.35                            0.600
    ## 4                             1.30                            0.300
    ##   Species: virginica (n=50) / N Species: virginica (n=50) / Mean
    ## 1                            50                            6.588
    ## 2                            50                            2.974
    ## 3                            50                            5.552
    ## 4                            50                               NA
    ##   Species: virginica (n=50) / sd Species: virginica (n=50) / Med
    ## 1                      0.6358796                            6.50
    ## 2                      0.3224966                            3.00
    ## 3                      0.5518947                            5.55
    ## 4                             NA                            2.00
    ##   Species: virginica (n=50) / IQR    tests / p                  tests / test
    ## 1                           0.675 1.505059e-28 oneway.test(., var.equal = F)
    ## 2                           0.375 4.492017e-17 oneway.test(., var.equal = T)
    ## 3                           0.775 4.803974e-29                  kruskal.test
    ## 4                           0.500 3.261796e-29                  kruskal.test

<br>

The result is a table containing a descriptive sub-table for each level
of the grouping factor (the statistical functions rules are applied to
each sub-table independently), with the statistical tests performed, and
their p values.

When displayed as a flat dataframe, the grouping header appears in each
variable name.

You can also see the grouping headers by inspecting the resulting
object, which is a nested list of dataframes, each dataframe being named
after the grouping factor and its levels (with sample size for each).

``` r
str(iris_by_Species)
```

    ## List of 5
    ##  $ Variables                 :'data.frame':  4 obs. of  1 variable:
    ##   ..$ Variables: chr [1:4] "Sepal.Length" "Sepal.Width" "Petal.Length" "Petal.Width"
    ##  $ Species: setosa (n=50)    :'data.frame':  4 obs. of  5 variables:
    ##   ..$ N   : int [1:4] 50 50 50 50
    ##   ..$ Mean: num [1:4] 5.01 3.43 NA NA
    ##   ..$ sd  : num [1:4] 0.352 0.379 NA NA
    ##   ..$ Med : num [1:4] 5 3.4 1.5 0.2
    ##   ..$ IQR : num [1:4] 0.4 0.475 0.175 0.1
    ##  $ Species: versicolor (n=50):'data.frame':  4 obs. of  5 variables:
    ##   ..$ N   : int [1:4] 50 50 50 50
    ##   ..$ Mean: num [1:4] 5.94 2.77 4.26 NA
    ##   ..$ sd  : num [1:4] 0.516 0.314 0.47 NA
    ##   ..$ Med : num [1:4] 5.9 2.8 4.35 1.3
    ##   ..$ IQR : num [1:4] 0.7 0.475 0.6 0.3
    ##  $ Species: virginica (n=50) :'data.frame':  4 obs. of  5 variables:
    ##   ..$ N   : int [1:4] 50 50 50 50
    ##   ..$ Mean: num [1:4] 6.59 2.97 5.55 NA
    ##   ..$ sd  : num [1:4] 0.636 0.322 0.552 NA
    ##   ..$ Med : num [1:4] 6.5 3 5.55 2
    ##   ..$ IQR : num [1:4] 0.675 0.375 0.775 0.5
    ##  $ tests                     :'data.frame':  4 obs. of  2 variables:
    ##   ..$ p   : num [1:4] 1.51e-28 4.49e-17 4.80e-29 3.26e-29

NEWS
Version 0.2.0

- Add support for purrr::map-like formulas for statistical and test functions
- "conditional" formulas are no longer supported. For example, replace `~ is.normal ~ mean | median` with `~ if (is.normal(.)) mean(.) else median(.)`

Version 0.1.9

- Fix in default options for datatable output to follow DT update
- Fix pander export for R ≥ 4.0.0

Version 0.1.8

- Code cleanup:
  - use RStudio style guidelines for all code and docs
  - use fewer tidyverse functions internally, drop `purrr` dependancy

Version 0.1.7

- Vignette and README with RStudio style guidelines

Version 0.1.6

- Correct way to re-export `group_by` and `%>%`

Version 0.1.4-5

- Documentation fix for group_by import (dplyr 0.8)
- Correct description for Dates

Version 0.1.3

- Added travis CI to repo
- Added CRAN badge to repo
- Corrected a typo in the README and vignette
- Bugfix for when a statistical function returns a warning AND an error
- Added NEWS file for future releases
- Modified default and auto statistics for less automation
- Added dummy column name for the variable names column, and make dplyr happy
- Enabled character and mixed columns
- Added CRAN mirror downloads badge to repo

Version 0.1.2

- Added a fix for a change in evaluation in rlang

Version 0.1.1

- Added a fix for changing arguments in fisher.test in upstream

Version 0.1.0

Initial release.


R/convenience_functions.R
#' Return the percentages for the levels of a factor
#'
#' Return a compatible vector of length nlevels(x) + 1
#' to print the percentages of each level of a factor
#'
#' @param x A factor
#' @export
#' @return A nlevels(x) + 1 length vector of percentages
percent <- function(x) {
  if (is.factor(x)) c(NA, summary(x, maxsum = Inf) / length(x)) * 100
  else              NA
}


#' Return the inter-quartile range
#'
#' Safe version of IQR for statify
#' @param x A vector
#' @return The IQR
#' @export
IQR <- function(x) {
  base::diff(stats::quantile(x, c(0.25, 0.75), na.rm = T))
}


#' Test if distribution is normal
#'
#' Test if distribution is normal.
#' The condition for normality is length > 30 and non-significant Shapiro-Wilks test with p > .1
#'
#' @param x A numerical vector
#' @export
#' @return A boolean
is.normal <- function(x) {
  if (!is.numeric(x)) F
  else if (length(stats::na.omit(x)) >= 30) tryCatch(stats::shapiro.test(x)$p.value > .1, error = function(e) F)
  else F
}


#' Fisher's Exact Test for Count Data
#'
#' Performs Fisher's exact test for testing the null of independence
#' of rows and columns in a contingency table with fixed marginals, or with a formula expression.
#'
#' If \code{x} is a matrix, it is taken as a two-dimensional contingency
#' table, and hence its entries should be nonnegative integers.
#' Otherwise, both \code{x} and \code{y} must be vectors of the same length.
#' Incomplete cases are removed, the vectors are coerced into factor
#' objects, and the contingency table is computed from these.
#'
#' For 2 by 2 cases, p-values are obtained directly using the
#' (central or non-central) hypergeometric distribution. Otherwise,
#' computations are based on a C version of the FORTRAN subroutine
#' FEXACT which implements the network developed by Mehta and Patel
#' (1986) and improved by Clarkson, Fan and Joe (1993).  The FORTRAN
#' code can be obtained from \url{http://www.netlib.org/toms/643}.
#' Note this fails (with an error message) when the entries of the
#' table are too large.  (It transposes the table if necessary so it
#' has no more rows than columns.  One constraint is that the product
#' of the row marginals be less than 2^31 - 1.)
#'
#' For 2 by 2 tables, the null of conditional independence is
#' equivalent to the hypothesis that the odds ratio equals one.
#' \code{Exact} inference can be based on observing that in general, given
#' all marginal totals fixed, the first element of the contingency
#' table has a non-central hypergeometric distribution with
#' non-centrality parameter given by the odds ratio (Fisher, 1935).
#' The alternative for a one-sided test is based on the odds ratio,
#' so \code{alternative = "greater"} is a test of the odds ratio being
#' bigger than \code{or}.
#'
#' Two-sided tests are based on the probabilities of the tables, and
#' take as \code{more extreme} all tables with probabilities less than or
#' equal to that of the observed table, the p-value being the sum of
#' such probabilities.
#'
#' For larger than 2 by 2 tables and \code{hybrid = TRUE}, asymptotic
#' chi-squared probabilities are only used if the ‘Cochran
#' conditions’ are satisfied, that is if no cell has count zero, and
#' more than 80% of the cells have counts at least 5: otherwise the
#' exact calculation is used.
#'
#' Simulation is done conditional on the row and column marginals,
#' and works only if the marginals are strictly positive.  (A C
#' translation of the algorithm of Patefield (1981) is used.)
#' @param x either a two-dimensional contingency table in matrix form, a factor object, or a formula of the form \code{lhs ~ rhs} where \code{lhs} and \code{rhs} are factors.
#' @param y a factor object; ignored if \code{x} is a matrix or a formula.
#' @param ... additional params to feed to original fisher.test
#' @inheritParams stats::fisher.test
#' @return A list with class \code{"htest"} containing the following components:
#'
#' p.value: the p-value of the test.
#'
#' conf.int: a confidence interval for the odds ratio.  Only present in
#'           the 2 by 2 case and if argument \code{conf.int = TRUE}.
#'
#' estimate: an estimate of the odds ratio.  Note that the _conditional_
#'           Maximum Likelihood Estimate (MLE) rather than the
#'           unconditional MLE (the sample odds ratio) is used.  Only
#'           present in the 2 by 2 case.
#'
#' null.value: the odds ratio under the null, \code{or}.  Only present in the 2
#'           by 2 case.
#'
#' alternative: a character string describing the alternative hypothesis.
#'
#' method: the character string \code{"Fisher's Exact Test for Count Data"}.
#'
#' data.name: a character string giving the names of the data.
#' @references
#' Agresti, A. (1990) _Categorical data analysis_.  New York: Wiley.
#' Pages 59-66.
#'
#' Agresti, A. (2002) _Categorical data analysis_. Second edition.
#' New York: Wiley.  Pages 91-101.
#'
#' Fisher, R. A. (1935) The logic of inductive inference.  _Journal
#' of the Royal Statistical Society Series A_ *98*, 39-54.
#'
#' Fisher, R. A. (1962) Confidence limits for a cross-product ratio.
#' _Australian Journal of Statistics_ *4*, 41.
#'
#' Fisher, R. A. (1970) _Statistical Methods for Research Workers._
#' Oliver & Boyd.
#'
#' Mehta, C. R. and Patel, N. R. (1986) Algorithm 643. FEXACT: A
#' Fortran subroutine for Fisher's exact test on unordered r*c
#' contingency tables.  _ACM Transactions on Mathematical Software_,
#' *12*, 154-161.
#'
#' Clarkson, D. B., Fan, Y. and Joe, H. (1993) A Remark on Algorithm
#' 643: FEXACT: An Algorithm for Performing Fisher's Exact Test in r
#' x c Contingency Tables.  _ACM Transactions on Mathematical
#' Software_, *19*, 484-488.
#'
#' Patefield, W. M. (1981) Algorithm AS159.  An efficient method of
#' generating r x c tables with given row and column totals.
#' _Applied Statistics_ *30*, 91-97.
#' @seealso
#' \code{\link{chisq.test}}
#'
#' \code{fisher.exact} in package \pkg{kexact2x2} for alternative
#' interpretations of two-sided tests and confidence intervals for 2
#' by 2 tables.
#' @examples
#' \dontrun{
#' ## Agresti (1990, p. 61f; 2002, p. 91) Fisher's Tea Drinker
#' ## A British woman claimed to be able to distinguish whether milk or
#' ##  tea was added to the cup first.  To test, she was given 8 cups of
#' ##  tea, in four of which milk was added first.  The null hypothesis
#' ##  is that there is no association between the true order of pouring
#' ##  and the woman's guess, the alternative that there is a positive
#' ##  association (that the odds ratio is greater than 1).
#' TeaTasting <-
#' matrix(c(3, 1, 1, 3),
#'        nrow = 2,
#'        dimnames = list(Guess = c("Milk", "Tea"),
#'                        Truth = c("Milk", "Tea")))
#' fisher.test(TeaTasting, alternative = "greater")
#' ## => p = 0.2429, association could not be established
#'
#' ## Fisher (1962, 1970), Criminal convictions of like-sex twins
#' Convictions <-
#' matrix(c(2, 10, 15, 3),
#'        nrow = 2,
#'        dimnames =
#'        list(c("Dizygotic", "Monozygotic"),
#'             c("Convicted", "Not convicted")))
#' Convictions
#' fisher.test(Convictions, alternative = "less")
#' fisher.test(Convictions, conf.int = FALSE)
#' fisher.test(Convictions, conf.level = 0.95)$conf.int
#' fisher.test(Convictions, conf.level = 0.99)$conf.int
#'
#' ## A r x c table  Agresti (2002, p. 57) Job Satisfaction
#' Job <- matrix(c(1,2,1,0, 3,3,6,1, 10,10,14,9, 6,7,12,11), 4, 4,
#' dimnames = list(income = c("< 15k", "15-25k", "25-40k", "> 40k"),
#'                 satisfaction = c("VeryD", "LittleD", "ModerateS", "VeryS")))
#' fisher.test(Job)
#' fisher.test(Job, simulate.p.value = TRUE, B = 1e5)
#'
#' ###
#' }
#' @export
fisher.test <- function(x, y, workspace, hybrid, control, or, alternative, conf.int, conf.level, simulate.p.value, B) {
  UseMethod("fisher.test")
}


#' @rdname fisher.test
fisher.test.default <- function(x, ...) {
  stats::fisher.test(x, ...)
}

#' @rdname fisher.test
fisher.test.formula <- function(x,
                                y = NULL,
                                workspace = 200000,
                                hybrid = F,
                                control = list(),
                                or = 1,
                                alternative = "two.sided",
                                conf.int = T,
                                conf.level = .95,
                                simulate.p.value = F,
                                B = 2000) {
  stats::fisher.test(x = eval(x[[2]], envir = parent.frame()),
                     y = eval(x[[3]], envir = parent.frame()),
                     workspace = workspace,
                     hybrid = hybrid,
                     control = control,
                     or = or,
                     alternative = alternative,
                     conf.int = conf.int,
                     conf.level = conf.level,
                     simulate.p.value = simulate.p.value,
                     B = B)
}


#' Pearson's Chi-squared Test for Count Data
#'
#' \code{chisq.test} performs chi-squared contingency table tests and goodness-of-fit tests, with an added method for formulas.
#'
#' If \code{x} is a matrix with one row or column, or if \code{x} is a vector
#' and \code{y} is not given, then a _goodness-of-fit test_ is performed
#' (\code{x} is treated as a one-dimensional contingency table).  The
#' entries of \code{x} must be non-negative integers.  In this case, the
#' hypothesis tested is whether the population probabilities equal
#' those in \code{p}, or are all equal if \code{p} is not given.
#'
#' If \code{x} is a matrix with at least two rows and columns, it is taken
#' as a two-dimensional contingency table: the entries of \code{x} must be
#' non-negative integers.  Otherwise, \code{x} and \code{y} must be vectors or
#' factors of the same length; cases with missing values are removed,
#' the objects are coerced to factors, and the contingency table is
#' computed from these.  Then Pearson's chi-squared test is performed
#' of the null hypothesis that the joint distribution of the cell
#' counts in a 2-dimensional contingency table is the product of the
#' row and column marginals.
#'
#' If \code{simulate.p.value} is \code{FALSE}, the p-value is computed from the
#' asymptotic chi-squared distribution of the test statistic;
#' continuity correction is only used in the 2-by-2 case (if
#' \code{correct} is \code{TRUE}, the default).  Otherwise the p-value is
#' computed for a Monte Carlo test (Hope, 1968) with \code{B} replicates.
#'
#' In the contingency table case simulation is done by random
#' sampling from the set of all contingency tables with given
#' marginals, and works only if the marginals are strictly positive.
#' Continuity correction is never used, and the statistic is quoted
#' without it.  Note that this is not the usual sampling situation
#' assumed for the chi-squared test but rather that for Fisher's
#' exact test.
#'
#' In the goodness-of-fit case simulation is done by random sampling
#' from the discrete distribution specified by \code{p}, each sample being
#' of size \code{n = sum(x)}.  This simulation is done in R and may be
#' slow.
#' @param x a numeric vector, or matrix, or formula of the form \code{lhs ~ rhs} where \code{lhs} and \code{rhs} are factors. \code{x} and \code{y} can also both be factors.
#' @param y a numeric vector; ignored if \code{x} is a matrix or a formula. If \code{x} is a factor, \code{y} should be a factor of the same length.
#' @inheritParams stats::chisq.test
#' @return A list with class \code{"htest"} containing the following components:
#' statistic: the value the chi-squared test statistic.
#'
#' parameter: the degrees of freedom of the approximate chi-squared
#'           distribution of the test statistic, \code{NA} if the p-value is
#'           computed by Monte Carlo simulation.
#'
#'  p.value: the p-value for the test.
#'
#'   method: a character string indicating the type of test performed, and
#'           whether Monte Carlo simulation or continuity correction was
#'           used.
#'
#' data.name: a character string giving the name(s) of the data.
#'
#' observed: the observed counts.
#'
#' expected: the expected counts under the null hypothesis.
#'
#' residuals: the Pearson residuals, ‘(observed - expected) /
#'           sqrt(expected)’.
#'
#'   stdres: standardized residuals, \code{(observed - expected) / sqrt(V)},
#'           where \code{V} is the residual cell variance (Agresti, 2007,
#'           section 2.4.5 for the case where \code{x} is a matrix, ‘n * p * (1
#'           - p)’ otherwise).
#' @source The code for Monte Carlo simulation is a C translation of the Fortran algorithm of Patefield (1981).
#' @references
#' Hope, A. C. A. (1968) A simplified Monte Carlo significance test
#' procedure.  _J. Roy, Statist. Soc. B_ *30*, 582-598.
#'
#' Patefield, W. M. (1981) Algorithm AS159.  An efficient method of
#' generating r x c tables with given row and column totals.
#' _Applied Statistics_ *30*, 91-97.
#'
#' Agresti, A. (2007) _An Introduction to Categorical Data Analysis,
#' 2nd ed._, New York: John Wiley & Sons.  Page 38.
#' @seealso For goodness-of-fit testing, notably of continuous distributions, \code{\link{ks.test}}.
#' @examples
#' \dontrun{
#' ## From Agresti(2007) p.39
#' M <- as.table(rbind(c(762, 327, 468), c(484, 239, 477)))
#' dimnames(M) <- list(gender = c("F", "M"),
#'                     party = c("Democrat","Independent", "Republican"))
#' (Xsq <- chisq.test(M))  # Prints test summary
#' Xsq$observed   # observed counts (same as M)
#' Xsq$expected   # expected counts under the null
#' Xsq$residuals  # Pearson residuals
#' Xsq$stdres     # standardized residuals
#'
#'
#' ## Effect of simulating p-values
#' x <- matrix(c(12, 5, 7, 7), ncol = 2)
#' chisq.test(x)$p.value           # 0.4233
#' chisq.test(x, simulate.p.value = TRUE, B = 10000)$p.value
#'                                 # around 0.29!
#'
#' ## Testing for population probabilities
#' ## Case A. Tabulated data
#' x <- c(A = 20, B = 15, C = 25)
#' chisq.test(x)
#' chisq.test(as.table(x))             # the same
#' x <- c(89,37,30,28,2)
#' p <- c(40,20,20,15,5)
#' try(
#' chisq.test(x, p = p)                # gives an error
#' )
#' chisq.test(x, p = p, rescale.p = TRUE)
#'                                 # works
#' p <- c(0.40,0.20,0.20,0.19,0.01)
#'                                 # Expected count in category 5
#'                                 # is 1.86 < 5 ==> chi square approx.
#' chisq.test(x, p = p)            #               maybe doubtful, but is ok!
#' chisq.test(x, p = p, simulate.p.value = TRUE)
#'
#' ## Case B. Raw data
#' x <- trunc(5 * runif(100))
#' chisq.test(table(x))            # NOT 'chisq.test(x)'!
#'
#' ###
#' }
#' @export
chisq.test <- function(x, y, correct, p, rescale.p, simulate.p.value, B) {
  UseMethod("chisq.test")
}


#' @rdname chisq.test
chisq.test.default <- stats::chisq.test

#' @rdname chisq.test

chisq.test.formula <- function(x,
                               y = NULL,
                               correct = T,
                               p = rep(1/length(x), length(x)),
                               rescale.p = F,
                               simulate.p.value = F,
                               B = 2000) {
  stats::chisq.test(x = eval(x[[2]], envir = parent.frame()),
                    y = eval(x[[3]], envir = parent.frame()),
                    correct = correct,
                    p = p,
                    rescale.p = rescale.p,
                    simulate.p.value = simulate.p.value,
                    B = B)
}


#' Wrapper for oneway.test(var.equal = T)
#'
#' @param formula An anova formula (\code{variable ~ grouping variable})
#' @seealso \code{\link{oneway.test}}
#' @export
ANOVA <- function(formula) {
  stats::oneway.test(formula, var.equal = T)
}


#' No test
#'
#' An empty test
#' @param formula A formula
no.test <- function(formula) {
  data.frame(p.value = NA)
}
